src/backend/EinsumNode.cpp

```diff
@@ -460,3 +511,32 @@ int64_t einsum_ir::backend::EinsumNode::num_ops( bool i_children ) {
 
   return l_num_ops;
 }
+
+void  einsum_ir::backend::EinsumNode::cancel_memory_reservation() {
+  m_num_mem_frees++;
+  if(m_num_mem_frees >= m_req_mem_frees && m_mem_id ){
+      m_memory->remove_reservation(m_mem_id);
+  }
+}
+
+
+void einsum_ir::backend::EinsumNode::calculate_memory_usage(){
+  // compile children
+  m_memory->increase_layer();
+  for( std::size_t l_ch = 0; l_ch < m_children.size(); l_ch++ ) {
+      m_children[m_exec_order[l_ch]] -> calculate_memory_usage();
+  }
+  m_memory->decrease_layer();
+
+  //allocate own mem
+  if( m_req_mem ) {
+    m_mem_id = m_memory->reserve_memory(m_req_mem);
+  }
+  //theoretical calculation
+
+  //deallocation of child memory
+  for( std::size_t l_ch = 0; l_ch < m_children.size(); l_ch++ ) {
+      m_children[l_ch]->cancel_memory_reservation();
+  }
+
+}
```

Comment:
Based on the name the method calculates the memory usage. But there seem to be other side effects, i.e., there are are also calls to ``reserve_memory`` and ``cancel_memory_reservation``.